Firestore/Source/API/FIRDocumentSnapshot.mm

```diff
@@ -187,48 +199,97 @@ - (FieldValueOptions)optionsForServerTimestampBehavior:
+      return value.boolean_value() ? @YES : @NO;
+    case FieldValue::Type::Integer:
+      return @(value.integer_value());
+    case FieldValue::Type::Double:
+      return @(value.double_value());
+    case FieldValue::Type::Timestamp:
+      return [self convertedTimestamp:value options:options];
+    case FieldValue::Type::ServerTimestamp:
+      return [self convertedServerTimetamp:value options:options];
+    case FieldValue::Type::String:
+      return util::WrapNSString(value.string_value());
+    case FieldValue::Type::Blob:
+      return MakeNSData(value.blob_value());
+    case FieldValue::Type::Reference:
+      return [self convertedReference:value];
+    case FieldValue::Type::GeoPoint:
+      return MakeFIRGeoPoint(value.geo_point_value());
+    case FieldValue::Type::Array:
+      return [self convertedArray:value.array_value() options:options];
+    case FieldValue::Type::Object:
+      return [self convertedObject:value.object_value() options:options];
+  }
+
+  UNREACHABLE();
+}
+
+- (id)convertedTimestamp:(const FieldValue &)value options:(const FieldValueOptions &)options {
+  FIRTimestamp *wrapped = MakeFIRTimestamp(value.timestamp_value());
+  if (options.timestamps_in_snapshots_enabled()) {
+    return wrapped;
   } else {
-    return [value valueWithOptions:options];
+    return [wrapped dateValue];
   }
 }
 
-- (NSDictionary<NSString *, id> *)convertedObject:(FSTObjectValue *)objectValue
-                                          options:(const FieldValueOptions &)options {
-  NSMutableDictionary *result = [NSMutableDictionary dictionary];
-  [objectValue.internalValue
-      enumerateKeysAndObjectsUsingBlock:^(NSString *key, FSTFieldValue *value, BOOL *stop) {
-        result[key] = [self convertedValue:value options:options];
-      }];
-  return result;
+- (id)convertedServerTimetamp:(const FieldValue &)value options:(const FieldValueOptions &)options {
+  const auto &sts = value.server_timestamp_value();
+  switch (options.server_timestamp_behavior()) {
+    case ServerTimestampBehavior::kNone:
+      return [NSNull null];
+    case ServerTimestampBehavior::kEstimate: {
+      FieldValue local_write_time = FieldValue::FromTimestamp(sts.local_write_time());
+      return [self convertedTimestamp:local_write_time options:options];
+    }
+    case ServerTimestampBehavior::kPrevious:
+      return sts.previous_value() ? [self convertedValue:*sts.previous_value() options:options]
+                                  : [NSNull null];
+    default:
+      HARD_FAIL("Unexpected server timestamp option: %s", options.server_timestamp_behavior());
+  }
 }
 
-- (NSArray<id> *)convertedArray:(FSTArrayValue *)arrayValue
+- (id)convertedReference:(const FieldValue &)value {
+  const auto &ref = value.reference_value();
+  const DatabaseId &refDatabase = ref.database_id();
+  const DatabaseId &database = _snapshot.firestore()->database_id();
+  if (refDatabase != database) {
+    // TODO(b/32073923): Log this as a proper warning.
```

Comment:
NB: This TODO was marked fixed in 2017.

(I'm not suggesting you fix it now since it was there before and this PR is already massive... just pointing it out. OTHO, grepping the code, I think this may be the only instance of NSLog in our production code. There's a few more in our test code though.)